File: leetcode/easy/ExcelSheetColumnTitle.py
class Solution:
    def convertToTitle(self, columnNumber: int) -> str:
        """
        Converts a 1-based integer column number to its corresponding Excel-style column title.

        This problem is analogous to converting a number to a base-26 system.
        However, standard base-26 uses digits 0-25, while Excel columns are 1-indexed
        (A=1, Z=26). This requires a slight modification to the standard base conversion algorithm.

        Algorithm:
        1. Loop as long as the columnNumber is greater than 0.
        2. In each iteration, we determine the rightmost character.
           - We subtract 1 from `columnNumber` to map the range [1, 26] to [0, 25].
           - We take the result modulo 26 (`(columnNumber - 1) % 26`) to get the
             0-indexed character code (0 for 'A', 25 for 'Z').
           - We convert this code to a character and append it to our result list.
        3. We then update `columnNumber` for the next iteration by integer dividing
           the decremented value by 26 (`(columnNumber - 1) // 26`). This effectively
           shifts to the next "digit" to the left.
        4. Since we build the string from right to left, we reverse it at the end.

        Example: columnNumber = 28
        - 1st loop: (28-1)%26=1 -> 'B'. result=['B']. columnNumber=(28-1)//26=1.
        - 2nd loop: (1-1)%26=0 -> 'A'. result=['B', 'A']. columnNumber=(1-1)//26=0.
        - Loop ends. Join reversed list: "AB".
        """
        result = []
        while columnNumber > 0:
            # Map 1-26 to 0-25 for modulo operation.
            remainder = (columnNumber - 1) % 26
            # Convert to character and append. 'A' is ASCII 65.
            result.append(chr(65 + remainder))
            # Update columnNumber for the next "digit".
            columnNumber = (columnNumber - 1) // 26

        # The result was built in reverse, so reverse it before joining.
        return "".join(reversed(result))

File: leetcode/easy/ExcelSheetColumnTitle.java
class Solution {
    /**
     * Converts a given positive integer (column number) to its corresponding
     * Excel column title.
     *
     * This problem is essentially a base-26 conversion, but with a crucial
     * difference: it's a 1-indexed system (A=1, B=2, ..., Z=26) rather than
     * a 0-indexed system (where 0 would typically map to 'A').
     *
     * Algorithm Breakdown:
     * 1. Initialize a `StringBuilder` to efficiently build the result string.
     * 2. Loop as long as `columnNumber` is greater than 0.
     * 3. In each iteration:
     *    a. Adjust `columnNumber` by subtracting 1. This transforms the 1-indexed
     *       system (1-26) into a 0-indexed system (0-25), which directly maps
     *       to character ASCII values ('A' to 'Z').
     *    b. Calculate the `index` of the current character by taking the modulo 26
     *       of the adjusted `columnNumber`.
     *    c. Convert this `index` back to its corresponding character ('A' + index)
     *       and append it to the `StringBuilder`.
     *    d. Update `columnNumber` by dividing the adjusted `columnNumber` by 26
     *       for the next iteration.
     * 4. After the loop, the `StringBuilder` contains the characters in reverse
     *    order (from least significant to most significant). Reverse it and
     *    convert it to a `String` to get the final Excel column title.
     *
     * Time Complexity: O(log_26 n), where n is the `columnNumber`. This is because
     *                  the `columnNumber` is divided by 26 in each iteration.
     * Space Complexity: O(log_26 n) for the `StringBuilder` to store the result.
     *
     * @param columnNumber The positive integer representing the Excel column number.
     * @return The corresponding Excel column title as a String.
     */
    public String convertToTitle(int columnNumber) {
        // Use StringBuilder for efficient string construction.
        StringBuilder resultBuilder = new StringBuilder();
        
        // Loop until the columnNumber becomes 0.
        while (columnNumber > 0) {
            // Step 3a: Adjust columnNumber by subtracting 1 to convert from 1-indexed (1-26)
            // to 0-indexed (0-25) for modulo and character mapping.
            columnNumber--;
            
            // Step 3b: Calculate the index of the current character (0-25).
            int index = columnNumber % 26;
            
            // Step 3c: Convert the index to its corresponding character ('A' + index)
            // and append it to the result.
            resultBuilder.append((char)('A' + index));
            
            // Step 3d: Update columnNumber for the next iteration by dividing by 26.
            columnNumber /= 26;
        }
        
        // Step 4: The characters were appended in reverse order, so reverse the StringBuilder
        // and convert it to a String.
        return resultBuilder.reverse().toString();
    }
}
